src/main/java/io/cdap/plugin/batch/joiner/Lookup.java

```diff
@@ -21,133 +21,272 @@
+
+      String outputFieldName = outputField.getAlias() == null ? outputField.getFieldName() : outputField.getAlias();
+      if (outputField.getFieldName().equals(outputFieldName)) {
+        // Record identity transform
+        if (perStageJoinKeys.get(outputField.getStageName()).contains(outputField.getFieldName())) {
+          // if the field is part of join key no need to emit the identity transform as it is already taken care
+          // by join
+          continue;
+        }
+        String operationName = String.format("Identity %s", stagedInputField);
+        FieldOperation identity = new FieldTransformOperation(operationName, IDENTITY_OPERATION_DESCRIPTION,
+                                                              Collections.singletonList(stagedInputField),
+                                                              outputFieldName);
+        operations.add(identity);
+        continue;
+      }
+
+      String operationName = String.format("Rename %s", stagedInputField);
+
+      FieldOperation transform = new FieldTransformOperation(operationName, RENAME_OPERATION_DESCRIPTION,
+                                                             Collections.singletonList(stagedInputField),
+                                                             outputFieldName);
+      operations.add(transform);
     }
-    StructuredRecord.Builder builder = StructuredRecord
-      .builder(Schema.recordOf("record", Schema.Field.of("id", fieldSchema)));
-    // depending on stage get the join key value
-    Object value = isLookupStage ? lookupValue : structuredRecord.get(config.getInputKeyField());
-    builder.set("id", value);
-    return Collections.singletonList(builder.build());
-  }
 
-  @Override
-  public JoinConfig getJoinConfig() throws Exception {
-    return joinConfig;
+    return operations;
   }
 
+  @Nullable
   @Override
-  public StructuredRecord merge(StructuredRecord joinKey, Iterable<JoinElement<StructuredRecord>> iterable)
-    throws Exception {
-    StructuredRecord.Builder builder = StructuredRecord.builder(Schema.parseJson(config.schema));
-    builder.set(config.getOutputField(), config.getDefaultValue());
-    final Iterator<JoinElement<StructuredRecord>> iterator = iterable.iterator();
-    while (iterator.hasNext()) {
-      final JoinElement<StructuredRecord> next = iterator.next();
-      if (next.getStageName().equals(config.getLookupDataset())) {
-        builder.set(config.getOutputField(), next.getInputRecord().get(config.getLookupValueField()));
+  public JoinDefinition define(AutoJoinerContext context) {
+    FailureCollector collector = context.getFailureCollector();
+    boolean hasUnknownInputSchema = context.getInputStages().values().stream().map(JoinStage::getSchema)
+      .anyMatch(Objects::isNull);
+    if (hasUnknownInputSchema) {
+      collector.addFailure("Schemas for all input stages must be provided.", null)
```

Comment:
Yes, it will fail validation - during the tests while switching the plugin to AutoJoiner we noticed that when input schemas are set as macro - on both validation and deployment "define" method receives input schemas as null. The values of macros for input schemas are not available during define stage. The plugin requires reading stage name and all columns of required stage in order to create the join definition. Is there any alternative on how to access input schemas in ``define`` method?

Reply:
In this case, output schema must be provided (either directly or via macro). https://github.com/data-integrations/lookup-transform/pull/2/files#diff-51ee9d0d309872eabf4a2fe0a788a36d2df8a20bf6d6b8989859f2854f72a731R219

Reply:
Please take a look at the joiner plugin, and see how it behaves when inputs are macros. https://github.com/cdapio/hydrator-plugins/blob/develop/core-plugins/src/main/java/io/cdap/plugin/batch/joiner/Joiner.java

Reply:
Joiner reads selected stages and fields from config (conf.getSelectedFields and conf.getJoinKeys methods) - user has the ability to select fields from input stage in the widget. In the case of this plugin user only provides name of the lookup dataset and and lookup field name in the widget - the rest of the fields (selected fields) are calculated from input schema. Please note that both stage name and fields are required as output schema does not contain information from which stage the field came from.

Reply:
Can you derive the selected fields from the provided output schema? Lookup key and value fields always come from the lookup dataset, everything else comes from the input dataset

Reply:
updated